app/src/main/java/org/mozilla/focus/browser/BrowserNavigationOverlay.kt

```diff
@@ -184,4 +261,166 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
+                setLayoutMarginParams(iconView, R.dimen.custom_home_tile_margin_value)
+            }
+        }.forceExhaustive
+
+        itemView.setOnClickListener {
+            onUrlEnteredListener.onNonTextInputUrlEntered(item.url)
+            TelemetryWrapper.homeTileClickEvent(item)
+        }
+
+        val tvWhiteColor = ContextCompat.getColor(holder.itemView.context, R.color.tv_white)
+        itemView.setOnFocusChangeListener { v, hasFocus ->
+            val backgroundResource: Int
+            val textColor: Int
+            if (hasFocus) {
+                backgroundResource = R.drawable.home_tile_title_focused_background
+                textColor = tvWhiteColor
+                menuButton.visibility = View.VISIBLE
+            } else {
+                backgroundResource = 0
+                textColor = Color.BLACK
+                menuButton.visibility = View.GONE
+            }
+            titleView.setBackgroundResource(backgroundResource)
+            titleView.setTextColor(textColor)
+        }
+    }
+
+    private fun setLayoutMarginParams(iconView: View, tileMarginValue: Int) {
+        val layoutMarginParams = iconView.layoutParams as ViewGroup.MarginLayoutParams
+        val marginValue = iconView.resources.getDimensionPixelSize(tileMarginValue)
+        layoutMarginParams.setMargins(marginValue, marginValue, marginValue, marginValue)
+        iconView.layoutParams = layoutMarginParams
+    }
+
+    fun getItemAtPosition(position: Int): HomeTile? {
+        if (position > -1 && position < itemCount) {
+            return tiles[position]
+        }
+        return null
+    }
+
+    fun addItemToFrontOfAdapter(homeTile: HomeTile) {
+        tiles.add(0, homeTile)
+        notifyItemInserted(0)
+    }
+
+    fun removeTileFromAdapter(url: String) {
+        for ((index, tile) in tiles.withIndex()) {
+            if (tile.url == url) {
+                removeItemAtPosition(index)
+                break
+            }
+        }
+    }
+
+    fun removeItemAtPosition(position: Int) {
+        if (position > -1 && position < itemCount) {
+            tiles.removeAt(position)
+            notifyItemRemoved(position)
+        }
+    }
+
+    override fun getItemCount() = tiles.size
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = TileViewHolder(
+            LayoutInflater.from(parent.context).inflate(R.layout.home_tile, parent, false)
+    )
+}
+
+private fun onBindBundledHomeTile(holder: TileViewHolder, tile: BundledHomeTile) = with (holder) {
+    val bitmap = BundledTilesManager.getInstance(itemView.context).loadImageFromPath(itemView.context, tile.imagePath)
+    iconView.setImageBitmap(bitmap)
+
+    titleView.text = tile.title
+}
+
+private fun onBindCustomHomeTile(uiLifecycleCancelJob: Job, holder: TileViewHolder, item: CustomHomeTile) = with (holder) {
+    launch(uiLifecycleCancelJob + UI, CoroutineStart.UNDISPATCHED) {
```

Comment:
\[detekt\] INFO: \[style/MaxLineLength\] Line detected that is longer than the defined maximum line length in the code style\.